-1

**Failed Tests**:   ClangBuild 
**Size**: This PR adds an extra 20KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d9e804/51179/summary.html
**COMMIT**: 9cb6cb1b1fca19cd2958acd81aaeb039f9c7d6f2
**CMSSW**: CMSSW_16_1_X_2026-02-08-0000/el8_amd64_gcc13
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/50071/51179/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Failed Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' /usr/bin/time -v scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Local Products Rules ..... started
>> Local Products Rules ..... done
>> Creating project symlinks
>> Entering Package TrackingTools/GeomPropagators
>> Compile sequence completed for CMSSW CMSSW_16_1_X_2026-02-08-0000
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 1
Command exited with non-zero status 1
	Command being timed: "scram build -k -j 32 COMPILER=llvm compile BUILD_LOG=yes"
	User time (seconds): 20.62
	System time (seconds): 5.25
	Percent of CPU this job got: 378%
</pre>

--- a/TrackingTools/GeomPropagators/interface/OptimalHelixPlaneCrossing.h
+++ b/TrackingTools/GeomPropagators/interface/OptimalHelixPlaneCrossing.h
@@ -5,48 +5,47 @@
 #include "TrackingTools/GeomPropagators/interface/HelixForwardPlaneCrossing.h"
 #include "TrackingTools/GeomPropagators/interface/HelixArbitraryPlaneCrossing.h"
 
-#include <memory>
+#include <variant>
 
 class OptimalHelixPlaneCrossing {
 public:
   using Base = HelixPlaneCrossing;
 
   template <typename... Args>
-  OptimalHelixPlaneCrossing(Plane const &plane, Args &&...args) {
+  explicit OptimalHelixPlaneCrossing(Plane const &plane, Args &&...args) {
     GlobalVector u = plane.normalVector();
     constexpr float small = 1.e-6;  // for orientation of planes
 
     if (std::abs(u.z()) < small) {
       // barrel plane:
       // instantiate HelixBarrelPlaneCrossing,
-      new (get()) HelixBarrelPlaneCrossingByCircle(args...);
+      mem_.emplace<HelixBarrelPlaneCrossingByCircle>(args...);
+      base_ = std::get_if<HelixBarrelPlaneCrossingByCircle>(&mem_);
     } else if ((std::abs(u.x()) < small) && (std::abs(u.y()) < small)) {
       // forward plane:
       // instantiate HelixForwardPlaneCrossing
-      new (get()) HelixForwardPlaneCrossing(args...);
+      mem_.emplace<HelixForwardPlaneCrossing>(args...);
+      base_ = std::get_if<HelixForwardPlaneCrossing>(&mem_);
     } else {
       // arbitrary plane:
       // instantiate HelixArbitraryPlaneCrossing
-      new (get()) HelixArbitraryPlaneCrossing(args...);
+      mem_.emplace<HelixArbitraryPlaneCrossing>(args...);
+      base_ = std::get_if<HelixArbitraryPlaneCrossing>(&mem_);
     }
   }
+  OptimalHelixPlaneCrossing() = delete;
+  OptimalHelixPlaneCrossing(const OptimalHelixPlaneCrossing &) = delete;
+  OptimalHelixPlaneCrossing &operator=(const OptimalHelixPlaneCrossing &) = delete;
+  OptimalHelixPlaneCrossing(OptimalHelixPlaneCrossing &&) = delete;
+  OptimalHelixPlaneCrossing &operator=(OptimalHelixPlaneCrossing &&) = delete;
 
-  ~OptimalHelixPlaneCrossing() { get()->~Base(); }
-
-  Base &operator*() { return *get(); }
-  Base const &operator*() const { return *get(); }
+  [[nodiscard]] Base &operator*() noexcept { return *base_; }
+  [[nodiscard]] Base const &operator*() const noexcept { return *base_; }
 
 private:
-  Base *get() { return (Base *)&mem; }
-  Base const *get() const { return (Base const *)&mem; }
-
-  union Tmp {
-    HelixBarrelPlaneCrossingByCircle a;
-    HelixForwardPlaneCrossing b;
-    HelixArbitraryPlaneCrossing c;
-  };
-  using aligned_union_t = typename std::aligned_storage<sizeof(Tmp), alignof(Tmp)>::type;
-  aligned_union_t mem;
+  std::variant<std::monostate, HelixBarrelPlaneCrossingByCircle, HelixForwardPlaneCrossing, HelixArbitraryPlaneCrossing>
+      mem_;
+  Base *base_;
 };
 
 #endif

